notarization_poller/src/notarization_poller/worker.py

```diff
@@ -0,0 +1,128 @@
+                    queue = Queue(
+                        options={
+                            "credentials": {"accessToken": self.config["taskcluster_access_token"], "clientId": self.config["taskcluster_client_id"]},
+                            "rootUrl": self.config["taskcluster_root_url"],
+                        },
+                        session=session,
+                    )
+                    new_tasks = await self._run_cancellable(claim_work(self.config, queue, num_tasks=num_tasks_to_claim))
+                self.last_claim_work = arrow.utcnow()
+                for claim_task in new_tasks.get("tasks", []):
+                    new_task = Task(self.config, claim_task)
+                    self.running_tasks.append(new_task)
+            await self.prune_running_tasks()
+            sleep_time = self.last_claim_work.timestamp + self.config["claim_work_interval"] - arrow.utcnow().timestamp
+            sleep_time > 0 and await self._run_cancellable(sleep(sleep_time))
+
+    async def prune_running_tasks(self):
+        """Prune any complete tasks from ``self.running_tasks``."""
+        for task in self.running_tasks:
+            if task.complete:
+                self.running_tasks.remove(task)
+
+    async def _run_cancellable(self, coroutine: typing.Awaitable):
+        if not self.is_cancelled:
+            self.future = asyncio.ensure_future(coroutine)
+            result = await self.future
+            self.future = None
+            return result
+
+    async def cancel(self, status=STATUSES["worker-shutdown"]):
+        """Cancel any running tasks."""
+        self.is_cancelled = True
+        self.future and self.future.cancel()
+        try:
+            await asyncio.wait([asyncio.ensure_future(task.stop(status=status)) for task in self.running_tasks])
+        except (asyncio.CancelledError, ValueError):
+            pass
+
+
+# main {{{1
+def main(event_loop=None):
+    """Notarization poller entry point: get everything set up, then enter the main loop.
+
+    Args:
+        event_loop (asyncio.BaseEventLoop, optional): the event loop to use.
+            If None, use ``asyncio.get_event_loop()``. Defaults to None.
+
+    """
+    event_loop = event_loop or asyncio.get_event_loop()
+    done = False
+    config = get_config_from_cmdln(sys.argv[1:])
+    update_logging_config(config)
+
+    log.info("Notarization poller starting up at {} UTC".format(arrow.utcnow().format()))
+    log.info("Worker FQDN: {}".format(socket.getfqdn()))
+    rm(config["work_dir"])
+    makedirs(config["work_dir"])
+    running_tasks = RunTasks(config)
+
+    async def _handle_sigterm():
+        log.info("SIGTERM received; shutting down")
+        nonlocal done
+        done = True
+        await running_tasks.cancel()
+
+    async def _handle_sigusr1():
+        """Stop accepting new tasks."""
+        log.info("SIGUSR1 received; no more tasks will be taken")
+        nonlocal done
+        done = True
```

Comment:
We cancel the task, and want to prevent the `while not done` from looping again. It's possible that while porting it from scriptworker I pruned too much or too little, but I'm not seeing the problem currently.

Reply:
I don't see anything that cancels a task in `_handle_sigusr1` (I do in `_handle_sigterm`). I also don't see anything in `RunTasks.invoke` that requires the loop here. I think this function should probably call something on `RunTasks` that causes it to break out of the loop, and the outer loop in `main` can be removed.

Reply:
Addressed in https://github.com/escapewindow/scriptworker-scripts/commit/bc304c110186ed77ea392408e1037e472142636a .